Add a { kind: "internal" } element descriptor to maintain init order

nicolo-ribaudo opened the pull request:
> _YAEDKL! (Yet another element descriptor kind lol)_

This PR fixes a bug which caused private and public fields not to be evaluated in order when there is a class decorator.
I introduced a new element descriptor, but developers can't interact with it:
1. It can't be removed;
1. It can't be duplicated;
1. It can only be created by the engine;
1. It can't be moved from one class to another.

Each "internal" element is associated to a private element or to an initializer, that class decorators can't access.

The descriptor of a class like this:
```js
class A {
  foo = 2;
  #bar = 3;
  baz = 4;
  #bax() {}
}
```

```jsonc
{ kind: "class",
  elements: [
    { kind: "field",
      key: "foo",
      placement: "own",
      descriptor: { /* ... */ },
    },
    { kind: "internal" },
    { kind: "field",
      key: "baz",
      placement: "own",
      descriptor: { /* ... */ },
    },
    { kind: "internal" },
  ],
}
```

You can find further discussion at https://github.com/tc39/proposal-decorators/issues/183 (also, please discuss in that issue and only comment in this PR to review the actual implementation, not the idea).

ljharb commented on spec.html:
```diff
@@ -150,32 +150,43 @@ <h1>The ElementDescriptor Specification Type</h1>
         The <dfn>ElementDescriptor</dfn> is a Record used to represent class elements at runtime.
         Values of the ElementDescriptor type are Record values whose fields are defined as by <emu-xref href="#table-element-descriptor-fields"></emu-xref>.
-        Unless otherwise specified, every field is always present.
+        Unless otherwise specified, every field is optional.
```
why this change?

nicolo-ribaudo replied:
Because if [[Kind]] is "internal" every field (except `[[Element]]`) is absent.

nicolo-ribaudo commented on spec.html:
```diff
@@ -822,30 +854,41 @@ <h1>FromElementDescriptor ( _element_ )</h1>
+        1. If _kind_ is `"internal"`,
+          1. If _internalElements_ is not set, throw a TypeError.
+          1. If _elementObject_ doesn't have an [[Element]] internal slot, throw a TypeError.
+          1. If _internalElements_ doesn't contain _elementObject_.[[Element]], throw a TypeError.
```
I kept these three checks separate because they are conceptually three different things:
1. `{ kind: "internal" }` was found in a descriptor outside of a class descriptor.
1. It was directly created by the user and not by the engine
1. It comes from another class.

littledan commented on spec.html:
```diff
@@ -736,16 +760,21 @@ <h1>DecorateConstructor ( _elements_, _decorators_ )</h1>
-          1. Let _privateElements_ be a List consisting of the ordered subsequence of _elements_ for each _element_ where _element_.[[Key]] is a Private Name.
-          1. Let _publicElements_ be a List consisting of the ordered subsequence of _elements_ for each _element_ where _element_.[[Key]] is not a Private Name.
-          1. Let _obj_ be FromClassDescriptor(_publicElements_).
+          1. Let _internalElements_ be an empty Set.
```
Nit: You can use set this way in web specs, but for the JS spec, better off to use a List...

littledan reviewed (commented):
I have to say, I'm not super excited about this solution. So many checks! They all make sense, but it ends up feeling somehow ad-hoc. Even with all those checks, this patch allows class decorators to observably permute private field declarations, which feels like it shouldn't be able to do (even though it's an edge case and unlikely to cause problems in practice). I'm not sure what a better solution is, though.

rbuckton commented:
This raises a few red flags for me as well. Maintaining initialization order is very important, as any change to the defined order could have unforseen consequences. I can see a few alternatives to this:

-  Make `elements` informational only. New elements could be added via an `extras`.
    - *This would maintain the internal order, but would prevent a class decorator from removing elements (maybe this is a good thing?).*
- Remove `elements` entirely until such a time as we have a mechanism to include private members in the list.
    - *This unfortunately would remove valuable information from class decoration.*
- Maintain an internal order for elements and use it regardless of changes to order in `elements`.
- Postpone class decorators.

nicolo-ribaudo commented:
I see a fifth option:
- Only allow elements to be added or removed to `elements`. It can be enforced in a way similar to the one used in this PR to disallow adding/removing internal elements.

> - Make elements informational only. New elements could be added via an extras.

Would it still allow to change the descriptors used inside elements? e.g.
```js
function decorator(desc) {
  desc.elements[0].writable = false;
}
```

If I had to choose one of @rbuckton options I would probably go with 2 or 4.

ljharb commented:
I have a use case for class decorators that requires me to be able to replace multiplr public fields, each with a different public field and a prototype method. Would this still be possible with any if these options?

nicolo-ribaudo commented:
With the option I proposed, yes: it only prevents reordering of _existing_ elements.

littledan commented:
I don't understand your fifth option: what would the enforcement be?

nicolo-ribaudo commented:
It disallows reordering existing elements. Sure, I can remove an element and create a new one with the same key and the same placement but in another position, but I it requires a big amount of work.

littledan commented:
I see, hmm. Seems like it would be valid, but it all somehow feels like too much. Let's consider @rbuckton's options 2 and 4 some more.

nicolo-ribaudo commented:
I'll update this PR for now, but I'm happy to discuss more.

nicolo-ribaudo committed:
Add a { kind: "internal" } element descriptor to maintain init order

Add some checks to `ClassDescriptor#elements`:
 - "internal" elements can't be added or removed.
 - Existing elements can't be reordered.

It is still possible to add new elements and remove existing
"visible" elements.

nicolo-ribaudo commented on spec.html:
```diff
@@ -737,16 +742,22 @@ <h1>DecorateConstructor ( _elements_, _decorators_ )</h1>
+              1. Add _element_ to _internalElements_.
+              1. Set _element_ to the ElementDescriptor { [[Kind]]: `"internal"`, [[Element]]: _element_ }.
+            1. Add _element_ to _visibleElements_.
+          1. Let _brand_ be a new unique Symbol value.
```
This is the part of this PR I feel worst about: I think that the behavior is what it should be, but the implementation feels too much hacky.
Here I used a symbol (an ecmascript value), which is then stored in an internal [[Brand]] slot, to check if the element descriptors come from this decorator invocation (and thus they shoudn't be reordered).

nicolo-ribaudo commented:
Another idea: instead of having the `elements` array, it can be an opaque object that can be only modified using two `map` and `push` methods, similarly to `children` in React which isn't really an array.
This would probably come after 2 or 4.

rbuckton commented:
An opaque object seems fairly complicated. Would `map` mutate the object in this case, or would you do `desc.elements = desc.elements.map(x => y)`? What if you wanted to map one descriptor to multiple descriptors? What if you wanted to insert a descriptor at the beginning (`unshift`) rather than the end (`push`)? What about enumerating or indexing into elements?

Another option would be to continue with the "internal" approach, but instead have a 1-1 mapping of private members to "internal" descriptors with an opaque identity like a symbol: `{ kind: "internal", placement: "static" | "prototype" | "own", id: Symbol() }`. The `id` is mapped to a private name internally, but doesn't expose the private name itself. We could then allow you to reorder/replace to your heart's content (footguns and all), and when applying the order we just look up the `id`. If we wanted to, we could throw an Error if you remove any "internal" descriptors (and ignore changes to the placement). 

This would also give us the ability to later support "trusted" class decorators. Depending on the trust model, we could either just _replace_ the "internal" descriptor with the actual descriptor, or provide an API to resolve an "internal" descriptor to the actual one. 

Finally, I have a small concern about the use of the word "internal" as some languages use that as a keyword to mean something slightly different. Perhaps just using "private" as the kind would be clearer and more consistent (i.e. "this is a _private_ member you don't have access to").

littledan commented:
@rbuckton I think I am missing something--how does your suggestion differ from the first patch set posted here?

littledan commented:
The name `kind: "private"` SGTM

rbuckton commented:
@littledan: Its not substantially different I suppose, though my suggestion depends less on object reference identity (i.e. `desc.elements[0] = { ...desc.elements[0] }` breaks the current proposal since the new object won't have the same internal slots, while it doesn't break in my suggestion since the important bit is the `id` symbol).

littledan commented:
I see, thanks love for explaining.

nicolo-ribaudo commented:
Oh right I didn't consider that. Your proposal is better.

littledan commented:
(FWIW I kind of prefer Nicolo's proposal, to not expose these internally-generated identifiers to JavaScript.)

nicolo-ribaudo commented:
It's a matter of deciding if the "internal" descriptors _are_ a reference to the existing descriptor or if _their data_ is that reference.

---

EDIT: In case it is needed, the original commit is c81ece9cba255405405030dcbf3e5ec0766c6c3d

Lodin commented:
@littledan, regarding [your comment](https://github.com/babel/babel/pull/9154#issuecomment-451525975): what would `elements` removing bring? Would I still have the ability to replace or remove unnecessary element? It is very important, because simple declaration of the property without any value already creates a property and two properties with same name are forbidden.

littledan commented:
Could you elaborate on your use case (or have you documented it somewhere)? Reusing the same private name will cause an error before the decorators even run.

Lodin commented:
@littledan Is it connected only with private names? I have an example with `connectedCallback` property.
```javascript
const customElement = ({elements, kind}) => {
  const superConnectedCallback = elements.find(({key}) => key === 'connectedCallback');

  return {
    elements: [
      ...elements.filter(({key}) => key === 'connectedCallback'),
      {
        descriptor: {
          value() {
            superConnectedCallback.descriptor.value.call(this);
            // do some rendering stuff
          },
        },
        key: 'connectedCallback',
        kind: 'method',
        placement: 'prototype',
      }
    ],
    kind,
  }
};
```
Example is kinda simplified, because I cannot guarantee that `connectedCallback` is always defined.

littledan commented:
OK, after some more thought, I think we should just revert the patch which filters out private class elements. I don't think any of us in the champion group liked that restriction anyway, but we thought we might be able to live with it. I'll post the extended rationale with the revert. For that reason, closing this PR.

littledan commented:
Thank you so much for doing the work to write all of this out, @nicolo-ribaudo ! It was very helpful in thinking through this problem space, and helping us realize that this path isn't viable. Talking abstractly, that wasn't so clear.
